bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/types/StateDescriptionFragmentImpl.java

```diff
@@ -0,0 +1,171 @@
+     *
+     * @param legacy the {@link StateDescription} to initialise from.
+     */
+    public StateDescriptionFragmentImpl(StateDescription legacy) {
+        this.minimum = legacy.getMinimum();
+        this.maximum = legacy.getMaximum();
+        this.step = legacy.getStep();
+        this.pattern = legacy.getPattern();
+        this.readOnly = Boolean.valueOf(legacy.isReadOnly());
+        if (legacy.getOptions() != null && !legacy.getOptions().isEmpty()) {
+            this.options = legacy.getOptions();
+        }
+    }
+
+    /**
+     * Copy constructor.
+     *
+     * @param source the source to copy from.
+     */
+    public StateDescriptionFragmentImpl(@NonNull StateDescriptionFragmentImpl source) {
+        this.minimum = source.getMinimum();
+        this.maximum = source.getMaximum();
+        this.step = source.getStep();
+        this.pattern = source.getPattern();
+        this.readOnly = source.isReadOnly();
+        this.options = source.getOptions();
+    }
+
+    @Override
+    public BigDecimal getMinimum() {
+        return minimum;
+    }
+
+    public void setMinimum(BigDecimal minimum) {
+        this.minimum = minimum;
+    }
+
+    @Override
+    public BigDecimal getMaximum() {
+        return maximum;
+    }
+
+    public void setMaximum(BigDecimal maximum) {
+        this.maximum = maximum;
+    }
+
+    @Override
+    public BigDecimal getStep() {
+        return step;
+    }
+
+    public void setStep(BigDecimal step) {
+        this.step = step;
+    }
+
+    @Override
+    public String getPattern() {
+        return pattern;
+    }
+
+    public void setPattern(String pattern) {
+        this.pattern = pattern;
+    }
+
+    @Override
+    public Boolean isReadOnly() {
+        return readOnly;
+    }
+
+    public void setReadOnly(Boolean readOnly) {
+        this.readOnly = readOnly;
+    }
+
+    @Override
+    public List<StateOption> getOptions() {
+        return options;
+    }
+
+    public void setOptions(List<StateOption> options) {
+        this.options = options;
+    }
+
+    @Override
+    public @Nullable StateDescription toStateDescription() {
+        if (minimum == null && maximum == null && step == null && readOnly == null && pattern == null
+                && options == null) {
+            return null;
+        }
+        final Boolean ro = readOnly;
+        return new StateDescription(minimum, maximum, step, pattern, ro == null ? false : ro.booleanValue(), options);
+    }
+
+    /**
+     * Merge the given {@link StateDescriptionFragment}. Set all unset ({@code null}) fields of this instance to the
+     * values from the given {@link StateDescriptionFragment}.
+     *
+     * @param fragment a {@link StateDescriptionFragment} this instance should merge in.
+     * @return this instance with the fields merged.
+     */
+    public StateDescriptionFragment merge(StateDescriptionFragment fragment) {
+        if (this.minimum == null) {
```

Comment:
I'm still surprised 😉 Wouldn't the intuitive way be 

```java
if (fragment.getMinimum() != null) {
  this.minimum = fragement.getMinimum();
}
```

?
At least _I_ would expect a "newest wins" strategy here (and for all others below). Which would be in compliance with 

> the higher [the rank], the more like it [the provided fragment] is going to win

in [StateDescriptionFragmentProvider.getRank()](https://github.com/eclipse/smarthome/pull/5238/files#diff-71d98e857913c7aaf3a5650f924b1a68R51).

Reply:
See also this test case: https://github.com/eclipse/smarthome/pull/5238/files#diff-df777383266487f98708b37c495e6960R175

Reply:
You're right:

https://github.com/eclipse/smarthome/blob/a31b18d194fe4310d29885db0dd924955989c7eb/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/service/StateDescriptionServiceImpl.java#L43-L49

Had to read it three times until it clicked... Wich brings me back to my most favorite "principle of least surprise": Shouldn't we also swap the comparison in order to sort it the other way round then?

Reply:
Aren't these implementation details? I don't mind turning both logics around, but I would like to have this one merged first. And then concentrate on this specific topic on a new PR. wdyt?